ethereum/evmtool/src/main/java/org/hyperledger/besu/evmtool/BlockchainTestSubCommand.java

```
@@ -231,89 +259,292 @@ private void traceTestSpecs(
+              final double timeMs = timeNs / 1_000_000.0;
+              parentCommand.out.printf(
+                  "Block %d (%s) Imported in %.2f ms (%.2f MGas/s)%n",
+                  block.getHeader().getNumber(), block.getHash(), timeMs, mGps);
+            } else {
+              parentCommand.out.printf(
+                  "Block %d (%s) Rejected (correctly)%n",
+                  block.getHeader().getNumber(), block.getHash());
+            }
+          }
+        } catch (final RLPException e) {
+          if (candidateBlock.isValid()) {
+            testPassed = false;
+            failureReason =
+                String.format(
+                    "Block %d (%s) RLP exception: %s",
+                    candidateBlock.getBlock().getHeader().getNumber(),
+                    candidateBlock.getBlock().getHash(),
+                    e.getMessage());
+            parentCommand.out.println(failureReason);
           }
         }
-      } catch (final RLPException e) {
-        if (candidateBlock.isValid()) {
-          testFailed = true;
-          failureReason =
-              String.format(
-                  "Block %d (%s) RLP exception: %s",
-                  candidateBlock.getBlock().getHeader().getNumber(),
-                  candidateBlock.getBlock().getHash(),
-                  e.getMessage());
-          parentCommand.out.println(failureReason);
+      }
+
+      if (!blockchain.getChainHeadHash().equals(spec.getLastBlockHash())) {
+        testPassed = false;
+        failureReason =
+            String.format(
+                "Chain header mismatch, have %s want %s",
+                blockchain.getChainHeadHash(), spec.getLastBlockHash());
+        parentCommand.out.printf(
+            "Chain header mismatch, have %s want %s - %s%n",
+            blockchain.getChainHeadHash(), spec.getLastBlockHash(), test);
+      } else {
+        parentCommand.out.println("Chain import successful - " + test);
+      }
+    } finally {
```

finally block adds unnecessary complexity IMO. In Java most exceptions are checked so you have to handle them, for a case like this where you are running tests if there's an unchecked one it's best to let them propagate and crash the app. What would be the benefit of getting a summary of results if there's a developer error?

---

Removed the finally block. Clean up happens at the end of normal execution flow.